File: py/sqrt.py
#!/usr/bin/python

"""
Notes:

(1)  Pre-optimized:
     
     It occurs to me, after writing this, that the function which searches for
     lower and upper bounds for the square root is not really necessary.  
     
     (a)  Using the same special case handling for numbers <= 1, we could just
          binary search bounds at [1, number) since, for any number > 1, that
          number's square root will be less than the number itself
     
     (b)  Even the special case handling for number <= 1 could be simplified.  
          The bounds could simply be [0, max(number, 1)].  It would be 
          worthwhile to write a "bare bones" implementation which omits many of
          the frills below and implements a simple binary search inwards.  
"""

import sys


FLOAT_EQUALS_DIFF = 0.0000001

def float_equals(a, b):
    """
    The greatest trick the devil ever played was not creating an "approximate equality" operator for floating point values
    
    Preconditions:  a, b must be floating point values
    
    Postconditions:  returns True if the a and b are "approximately equal"
    """
    
    return abs(a - b) < FLOAT_EQUALS_DIFF


def find_square_root_bounds(number, starting_lower_bound=2.0):
    """
    Given a search value and a starting lower bound, finds a lower and upper bound for the square root of that value
    
    Preconditions: number must be a positive floating point value
    
    Postconditions: returns a pair of floating point values whose squares are
                    respectively LTE and GTE the arg "number"
    """
    
    current_bound = starting_lower_bound
    while current_bound * current_bound < number:
        current_bound *= 2.0
    return (current_bound / 2.0, current_bound)


def main(argv):
    if len(argv) < 2:
        sys.stderr.write("Usage: %s <number>" % (argv[0],))
        return 1

    number = float(argv[1])
    accuracy = 0.001
    assert FLOAT_EQUALS_DIFF < accuracy

    print "Estimating square root of %g with minimum accuracy of %g" % (number, accuracy)

    sqrt = None
    
    # special case: negative numbers or exact matches for 0 or 1
    if float_equals(number, 0.0):
        sqrt = 0.0
    elif number < 0.0:
        sqrt = None
    elif float_equals(number, 1.0):
        sqrt = 1.0
    
    # let's actually compute an estimated square root
    else:
        if number < 1.0:
            lower_bound = 0.0
            upper_bound = 1.0
        else:
            lower_bound, upper_bound = find_square_root_bounds(number)
        
        # special case: early termination on exact matches
        if float_equals(lower_bound * lower_bound, number):
            sqrt = lower_bound
        elif float_equals(upper_bound * upper_bound, number):
            sqrt = upper_bound
        else:
            # binary search within bounds for an accurate estimate
            while upper_bound - lower_bound > accuracy:
                midpoint = (upper_bound + lower_bound) / 2.0
                
                # special case: early termination on exact matches
                if float_equals(midpoint * midpoint, number):
                    lower_bound = midpoint
                    upper_bound = midpoint
                    break
                elif midpoint * midpoint < number:
                    lower_bound = midpoint
                else:
                    upper_bound = midpoint
        
            sqrt = lower_bound
    
    if sqrt is not None:
        print "Calculated estimated square root:\n  %g" % (sqrt)
    else:
        print "Cannot calculate square root for a negative number.  What sort of imaginary number machine do you think this is?"


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: cpp/sqrt.cpp
/* Re-implementing the square root estimator in c++
 */

#include <stdio.h>
#include <math.h>

double DOUBLE_ACCURACY = 0.000000001;

double ACCURACY_THRESHOLD = 0.00001;

/* Estimate equality on two double values */
bool double_equals(double a, double b) {
    return fabs(a - b) < DOUBLE_ACCURACY;
}

/* Estimate the square root of the input within accuracy of ACCURACY_THRESHOLD.
   On negative input, "-1.0" is returned as the "error".  
*/
double square_root(const double number) {
    // error check: number must be greater than 1
    double retval;
    if (double_equals(number, 0.0)) {
        retval = 0.0;
    }
    else if (number < 0.0) {
        retval = -1.0;
    }
    else {
        double lower_bound = 0.0;
        double upper_bound = fmax(number, 1.0);
        double midpoint;

        while (upper_bound - lower_bound > ACCURACY_THRESHOLD) {
            midpoint = (upper_bound + lower_bound) / 2.0;
            if (midpoint * midpoint > number) {
                upper_bound = midpoint;
            }
            else {
                lower_bound = midpoint;
            }
        }

        retval = lower_bound;
    }

    return retval;
}

/* Input: one double value
   Output: None

   Prints messages to stdio regarding the estimated
   square root of the input value
 */
int main(const int argc, const char* argv[])
{
    double input;
    // TODO:  Add error checking for number of arguments
    int sscanf_out = sscanf(argv[1], "%lg", &input);
    // TODO:  Add error checking for translation errors

    printf("Calculating square root of\n%lg\n", input);

    double est_sqrt = square_root(input);

    printf("Estimate:\n%lg\n", est_sqrt);

    return 0;
}
